tests/csapi/room_relations_test.go

```diff
@@ -205,3 +206,91 @@ func TestRelationsPagination(t *testing.T) {
 		},
 	})
 }
+
+func TestRelationsPaginationSync(t *testing.T) {
+  runtime.SkipIf(t, runtime.Dendrite) // FIXME: https://github.com/matrix-org/dendrite/issues/2944
+
+	deployment := Deploy(t, b.BlueprintAlice)
+	defer deployment.Destroy(t)
+
+	alice := deployment.Client(t, "hs1", "@alice:hs1")
+	roomID := alice.CreateRoom(t, map[string]interface{}{"preset": "public_chat"})
+	_, token := alice.MustSync(t, client.SyncReq{})
+
+	res := alice.MustDoFunc(t, "PUT", []string{"_matrix", "client", "v3", "rooms", roomID, "send", "m.room.message", "txn-1"}, client.WithJSONBody(t, map[string]interface{}{
+		"msgtype": "m.text",
+		"body":    "root",
+	}))
+	rootEventID := client.GetJSONFieldStr(t, client.ParseJSON(t, res), "event_id")
+
+	// Create some related events.
+	event_id := ""
+	for i := 0; i < 5; i++ {
+		res = alice.MustDoFunc(t, "PUT", []string{"_matrix", "client", "v3", "rooms", roomID, "send", "m.room.message", fmt.Sprintf("txn-%d", 2+i)}, client.WithJSONBody(t, map[string]interface{}{
+			"msgtype": "m.text",
+			"body":    fmt.Sprintf("reply %d", i),
+			"m.relates_to": map[string]interface{}{
+				"event_id": rootEventID,
+				"rel_type": "m.thread",
+			},
+		}))
+		event_id = client.GetJSONFieldStr(t, client.ParseJSON(t, res), "event_id")
+	}
+
+	// Sync and keep the token.
+	nextBatch := alice.MustSyncUntil(t, client.SyncReq{Since: token}, client.SyncTimelineHas(roomID, func(r gjson.Result) bool {
```

Comment:
There is a `SyncTimelineHasEventID` helper which might save some typing:

https://github.com/matrix-org/complement/blob/70f58acc4576daddaba968a8a29db55b04418c5d/internal/client/client.go#L793-L798

Reply:
Huh, was that always there?

Reply:
No, I split it out of `SendEventSynced` recently, in #588